keras/losses/loss.py

```diff
@@ -126,48 +109,38 @@ def reduce_weighted_values(
     values = ops.convert_to_tensor(values, dtype=dtype)
     if sample_weight is not None:
         sample_weight = ops.convert_to_tensor(sample_weight, dtype=dtype)
-    if mask is not None:
-        mask = ops.convert_to_tensor(mask, dtype=dtype)
+        sample_weight, values = squeeze_to_same_rank(sample_weight, values)
+        values = values * sample_weight
 
-    # Merge mask and sample weight into sample weight.
-    sample_weight = apply_mask(
-        sample_weight, mask, dtype=values.dtype, reduction=reduction
-    )
+    if mask is not None:
+        mask = ops.cast(mask, "bool")
+        mask, values = squeeze_to_same_rank(mask, values)
+        values = ops.where(mask, values, ops.zeros_like(values))
 
-    if sample_weight is not None:
-        sample_weight = ops.cast(sample_weight, values.dtype)
-        # Update dimensions of `sample_weight` to match `losses`.
-        values, sample_weight = squeeze_to_same_rank(values, sample_weight)
-        values = values * sample_weight
+    if reduction is None or reduction == "none":
+        return values
 
-    # Apply reduction function to the individual weighted losses.
-    loss = reduce_values(values, reduction)
-    return loss
+    if reduction == "sum":
+        return ops.sum(values)
 
+    if reduction == "sum_over_batch_size":
+        if mask is None:
+            # batch_size is the total number of elements
+            return ops.mean(values)
+        batch_size = ops.count_nonzero(mask)
+        values_sum = ops.sum(values)
+        # safe divide
+        return ops.cond(
```

Comment:
This will add significant overhead to the computation. Is there a better way that doesn't involve a cond?

Reply:
There are plenty of ways to do essentially this, but is a `cond` on scalars that expensive? What about a `where`? They seemed the clearest, but other options would be:
- use `batch_size = maximum(batch_size, 1)` - should give identical results (batch size is an integer, and when it's 0 the numerator should be zero)
- use `batch_size = batch_size + epsilon` - closer to the current implementation, though IMO wrong.